### PR TITLE
Supports disabling SSL certificate verification

### PR DESCRIPTION
**Pull request by ElasticCoder**
See [Issue 42](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/42). 

This PR will add the DisableSslVerification property to the CosmosConfiguration. It then gets set on the ConnectionPolicy (this property was present on the the Java and Python SDKs).

Finally the HttpRequestMessageHandler is modified to take the ConnectionPolicy as a constructor parameter. If the DisableSslVerification is true, then a certificate bypass function is added to the HttpClientHandler.

This was tested on a microsoft/dotnet:runtime image running on a Linux container. By setting 
DisableSslVerification to true, and DisableSslVerification ConnectionMode = ConnectionMode.Gateway, then the container can connect to the emulator running on the host machine.

Without these settings, the container complains about unsigned certificates with the "The remote certificate is invalid according to the validation procedure" error. The exception is thrown even if you install the self-signed certificate into the ca-certificates store. This is due to the way the .NET core code interops with the cURL/OpenSSL libraries. (They return error 18 for a self-signed certificate). This does not occur with Windows containers.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs [hunk 1]
```diff
@@ -6785,11 +6785,22 @@ private class HttpRequestMessageHandler : DelegatingHandler
...
+            public HttpRequestMessageHandler(EventHandler<SendingRequestEventArgs> sendingRequest, EventHandler<ReceivedResponseEventArgs> receivedResponse, ConnectionPolicy connectionPolicy)
             {
                 this.sendingRequest = sendingRequest;
                 this.receivedResponse = receivedResponse;
-                InnerHandler = new HttpClientHandler();
+
+                var clientHandler = new HttpClientHandler();
+                if (connectionPolicy.DisableSslVerification)
+                {
+                    clientHandler.ClientCertificateOptions = ClientCertificateOption.Manual;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/DocumentClient.cs]: Is Manul must? 
Will automatic and 'ServerCertificateCustomValidationCallback' override work?

**Comment by kirankumarkolli** [comment]
This failure is supposed to be only for emulator right? 
Also did you get chance to test in Direct (with TCP) mode?

**Comment by kirankumarkolli** [comment]
@ElasticCoder thanks a-lot for proactive PR. This Will benefit community a-lot.

**Comment by ElasticCoder** [comment]
Yes, this failure is for connecting to the emulator, and is only for Linux clients too. Windows clients honour the self-signed certificate installed in the Trusted Certification Authorities whereas the OpenSSL code returns an error code which is treated as fatal, even if the certificate is installed in ca-certificates. 

I will validate these changes against direct mode.

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs (hunk 1 above)
Comment (ElasticCoder) [reply Microsoft.Azure.Cosmos/src/DocumentClient.cs]: You are correct. Manual is not required, I have retested and it works without it. In fact, [Manual ](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.clientcertificateoption?view=netcore-2.1)is the default value.

**Comment by ElasticCoder** [comment]
> This failure is supposed to be only for emulator right?
> Also did you get chance to test in Direct (with TCP) mode?

I have tested it now with Direct TCP mode, and it doesn't work.

Now, I don't think this is a failure of the code or the change that I have made. Instead I think it is because the container is unable to resolve the emulator address properly when operating in TCP mode. I had this before I even started on the PR.

The error I get is:
```
CosmosRequestException;StatusCode=ServiceUnavailable;SubStatusCode=0;
ActivityId=5447f261-5025-4ade-a60f-b5be4b78d0c4;
RequestCharge=0;
Message=Response status code does not indicate success: 503 Substatus: 0 Reason: (Service is currently unavailable.ActivityId: 5447f261-5025-4ade-a60f-b5be4b78d0c4, RequestStartTime: 2019-03-04T19:27:16.5685846Z, 
Number of regions attempted: 2
ResponseTime: 2019-03-04T19:27:16.7855673Z, 
StoreReadResult: StorePhysicalAddress: rntbd://10.0.75.1:10251/, 
LSN: -1, 
GlobalCommittedLsn: -1, PartitionKeyRangeId: , IsValid: False, StatusCode: 410, IsGone: True, IsNotFound: False, IsInvalidPartition: False, 
RequestCharge: 0, ItemLSN: -1, 
SessionToken: , 
ResourceType: Database, 
OperationType: ReadResponseTime: 2019-03-04T19:27:16.7860857Z, 
StoreReadResult: StorePhysicalAddress: rntbd://10.0.75.1:10251/, 
```

I will need to debug through the code even deeper to find the source of the problem. For now, HTTPS and the Gateway mode with the DisableSslVerification property set to true helps to work around that.

**Comment by ElasticCoder** [comment]
[This issue ](https://github.com/Azure/azure-cosmos-dotnet-v2/issues/194)sounds relevant. I'll keep exploring.

**Comment by kirankumarkolli** [comment]
Thanks @ElasticCoder . Sometime tomorrow will also spend my-self to look through Direct code paths.

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs (hunk 1 above)
Comment (ElasticCoder) [reply Microsoft.Azure.Cosmos/src/DocumentClient.cs]: Added commit f0f7d which removes the ClientCertificateOption.Manual line.

**Comment by ElasticCoder** [comment]
A brief update: I have noticed that the V3 SDK only allows TCP with Direct mode, whereas [this article ](https://docs.microsoft.com/en-us/azure/cosmos-db/performance-tips)mentions that HTTPS and TCP should be permitted with Direct. Is that by design, or an oversight?

**Comment by ElasticCoder** [comment]
Pause this PR for now, I am going to look at an alternative option which is to inject an HttpClientHandler into the DocumentClient, This will allow the consumer to override SSL handling and also configure for working behind a proxy.

**Comment by ElasticCoder** [comment]
> Thanks @ElasticCoder . Sometime tomorrow will also spend my-self to look through Direct code paths.

I've been doing a lot of testing trying to get separate Linux machines on the network to connect to the emulator so I can step through the code. I've raised a [feature request on the emulator](https://feedback.azure.com/forums/263030-azure-cosmos-db/suggestions/37070941-select-network-interface-for-cosmos-db-emulator-to). This may be the cause of the "503 Gone" exception and the "A task was canceled" errors that people have been encountering.

**Comment by ElasticCoder** [comment]
Closing this pull request as it has been addressed in #74.

**Pull request closed without merging**